python/revng/internal/api/synchronizer.py

```diff
@@ -0,0 +1,309 @@
+
+
+class BotocoreSession(botocore.session.Session):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.credentials = botocore.credentials.Credentials("", "")
+
+    def get_credentials(self):
+        return self.credentials
+
+    def set_credentials(self, access_key: str, secret_key: str):  # type: ignore
+        self.credentials.access_key = access_key
+        self.credentials.secret_key = secret_key
+
+
+@dataclass(frozen=True)
+class FileData:
+    s3_filename: str
+    index: int = field(default=0)
+    encoding: str = field(default="None")
+
+
+@contextmanager
+def with_event(ev: Event):
+    """Exception/return-safe way of set()-ing an Event. When the __exit__
+    method of the context manager is called the event will be set"""
+    yield None
+    ev.set()
+
+
+def gather_bools(future_list: Sequence[Future[bool]], exception_message: str) -> bool:
+    """Similar to asyncio.gather for a list of Future[bool]. Will gracefully
+    manage multiple exceptions by returning an ExceptionGroup. The returned
+    result is True if all the awaited booleans are true and false otherwise."""
+    exceptions: list[Exception] = []
+    result = True
+    for future in future_list:
+        try:
+            if not future.result():
+                result = False
+        except Exception as ex:
+            exceptions.append(ex)
+
+    if len(exceptions) > 0:
+        raise ExceptionGroup(exception_message, exceptions)
+
+    return result
+
+
+class S3Synchronizer(Synchronizer):
+    def __init__(self, url: str):
+        self.credentials: str | None = None
+        # Url format is:
+        # s3(s)://<username>:<password>@<region>+<host:port>/<bucket name>/<path>
+        match_obj = re.match(
+            r"^(?P<proto>s3(|s))://(?P<username>[^:]*):(?P<password>[^@]*)@"
+            r"(?P<region>[^+]*)\+(?P<host>[^/]*)/(?P<bucket>[^/]+)(?P<path>/.*)?$",
+            url,
+        )
+        if match_obj is None:
+            raise ValueError("S3 endpoint invalid")
+
+        self.botocore_session = BotocoreSession()
+        self.botocore_session.set_credentials(match_obj["username"], match_obj["password"])
+        config = botocore.config.Config(s3={"addressing_style": "path"})
+        session = boto3.session.Session(botocore_session=self.botocore_session)
+        proto = "https" if match_obj["proto"] == "s3s" else "http"
+        self.client = session.client(
+            "s3",
+            endpoint_url=f"{proto}://{match_obj['host']}",
+            region_name=match_obj["region"],
+            config=config,
+        )
+        self.bucket = match_obj["bucket"]
+        if match_obj["path"] is not None:
+            self.path = match_obj["path"][1:]
+        else:
+            self.path = ""
+
+        self.last_save_event: Event | None = None
+        self.last_save_event_lock = Lock()
+        self.last_file_list: dict[str, FileData] = {}
+        self.last_file_list_lock = Lock()
```

Comment:
1. Add comment about the fact that this class needs to be thread-safe
2. Clearly divide "read-only" fields from fields that need to be updated.
3. Rename `*_lock` to `lock` and state that you need to acquire it before changing non-read-only fields.

Reply:
Done 1-3